[exporter/datadog] Data race reported in Windows tests

### Component(s)

exporter/datadog

### Describe the issue you're reporting

See https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8208921027/job/22453372508

```
==================
WARNING: DATA RACE
Read at 0x00c000d98250 by goroutine 1258:
  runtime.chansend()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/runtime/chan.go:160 +0x0
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload.func1()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:267 +0x4a5

Previous write at 0x00c000d98250 by goroutine 1234:
  runtime.closechan()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/runtime/chan.go:357 +0x0
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).createMetricsExporter.func2()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:367 +0xe9
  go.opentelemetry.io/collector/component.ShutdownFunc.Shutdown()
      C:/Users/runneradmin/go/pkg/mod/go.opentelemetry.io/collector/component@v0.96.1-0.20240306115632-b2[693](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8208921027/job/22453372508#step:7:694)620eff6/component.go:86 +0x118
  go.opentelemetry.io/collector/exporter/exporterhelper.(*baseExporter).Shutdown()
      C:/Users/runneradmin/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.96.1-0.20240306115632-b2693620eff6/exporterhelper/common.go:243 +0x55
  go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsExporter).Shutdown()
      <autogenerated>:1 +0x5b
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle.func4()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:63 +0x1ab
  testing.tRunner()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x44

Goroutine 1258 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:246 +0x5d
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).createMetricsExporter()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:311 +0xa84
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).createMetricsExporter-fm()
      <autogenerated>:1 +0xc4
  go.opentelemetry.io/collector/exporter.CreateMetricsFunc.CreateMetricsExporter()
      C:/Users/runneradmin/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.96.1-0.20240306115632-b2693620eff6/exporter.go:113 +0xd6
  go.opentelemetry.io/collector/exporter.(*factory).CreateMetricsExporter()
      <autogenerated>:1 +0x29
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle.func2()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:40 +0xc1
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle.func4()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:61 +0x155
  testing.tRunner()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x44

Goroutine 1234 (finished) created at:
  testing.(*T).Run()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x845
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:60 +0x5e4
  testing.tRunner()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
ERROR rerun aborted because previous run had a suspected panic and some test may not have run
make[2]: *** [../../Makefile.Common:126: test] Error 3
make[1]: *** [Makefile:165: exporter/datadogexporter] Error 2
make[1]: *** Waiting for unfinished jobs....
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x44
==================
panic: send on closed channel

goroutine 1282 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload.func1()
	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:267 +0x4a6
created by github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload in goroutine 1210
	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:246 +0x5e

=== FAIL: . TestComponentLifecycle/metrics-lifecycle (unknown)
==================
WARNING: DATA RACE
Read at 0x00c000d98250 by goroutine 1258:
  runtime.chansend()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/runtime/chan.go:160 +0x0
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload.func1()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:267 +0x4a5

Previous write at 0x00c000d98250 by goroutine 1234:
  runtime.closechan()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/runtime/chan.go:357 +0x0
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).createMetricsExporter.func2()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:367 +0xe9
  go.opentelemetry.io/collector/component.ShutdownFunc.Shutdown()
      C:/Users/runneradmin/go/pkg/mod/go.opentelemetry.io/collector/component@v0.96.1-0.20240306115632-b2693620eff6/component.go:86 +0x118
  go.opentelemetry.io/collector/exporter/exporterhelper.(*baseExporter).Shutdown()
      C:/Users/runneradmin/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.96.1-0.20240306115632-b2693620eff6/exporterhelper/common.go:243 +0x55
  go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsExporter).Shutdown()
      <autogenerated>:1 +0x5b
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle.func4()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:63 +0x1ab
  testing.tRunner()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x44

Goroutine 1258 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:246 +0x5d
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).createMetricsExporter()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:311 +0xa84
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).createMetricsExporter-fm()
      <autogenerated>:1 +0xc4
  go.opentelemetry.io/collector/exporter.CreateMetricsFunc.CreateMetricsExporter()
      C:/Users/runneradmin/go/pkg/mod/go.opentelemetry.io/collector/exporter@v0.96.1-0.20240306115632-b2693620eff6/exporter.go:113 +0xd6
  go.opentelemetry.io/collector/exporter.(*factory).CreateMetricsExporter()
      <autogenerated>:1 +0x29
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle.func2()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:40 +0xc1
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle.func4()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:61 +0x155
  testing.tRunner()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x44

Goroutine 1234 (finished) created at:
  testing.(*T).Run()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x845
  github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.TestComponentLifecycle()
      D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/generated_component_test.go:60 +0x5e4
  testing.tRunner()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1595 +0x261
  testing.(*T).Run.func1()
      C:/hostedtoolcache/windows/go/1.21.8/x64/src/testing/testing.go:1648 +0x44
==================
panic: send on closed channel

goroutine 1282 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload.func1()
	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:267 +0x4a6
created by github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*factory).consumeStatsPayload in goroutine 1210
	D:/a/opentelemetry-collector-contrib/opentelemetry-collector-contrib/exporter/datadogexporter/factory.go:246 +0x5e
```

--- a/.chloggen/ddog-data-race.yaml
+++ b/.chloggen/ddog-data-race.yaml
@@ -1,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: datadogexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix data race in metrics exporter shutdown
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [31663]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
+
+# If your change doesn't affect end users or the exported elements of any package,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+# Optional: The change log or logs in which this entry should be included.
+# e.g. '[user]' or '[user, api]'
+# Include 'user' if the change is relevant to end users.
+# Include 'api' if there is a change to a library API.
+# Default: '[user]'
+change_logs: []

--- a/exporter/datadogexporter/factory.go
+++ b/exporter/datadogexporter/factory.go
@@ -355,7 +355,8 @@ func (f *factory) createMetricsExporter(
 		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: true}),
 		exporterhelper.WithQueue(cfg.QueueSettings),
 		exporterhelper.WithShutdown(func(context.Context) error {
-			cancel()
+			cancel()    // first cancel context
+			f.wg.Wait() // then wait for shutdown
 			f.StopReporter()
 			statsWriter.Stop()
 			if statsIn != nil {

